nanowar.ts

```diff
@@ -0,0 +1,314 @@
+    try {
+        if (numberOfTroops === 0) return [];
+        let lines = input.split("\n");
+        let troops: UserStep = [];
+        let fromTo = new Set<string>();
+        for (let i = 0; i < numberOfTroops; i++) {
+            let [ from, to, size ] = lines[i].split(" ").map(x => parseInt(x));
+            if (state.tick.planets.length < from || state.tick.planets.length < to) {
+                return { error: "Invalid planet id" };  // TODO: do not punish so strongly. Just ignore one line if it is invalid, not the whole step.
+            }
+            let planetPlayer = state.tick.planets[from].player;
+            if (planetPlayer === null) {
+                return { error: "Invalid Planet! Planet is not owned by any player" };
+            }
+            if (planetPlayer.id !== playerID) {
+                return { error: "Invalid planet! Planet is owned by other player." }; // TODO: insert planet number and owner number
+            }
+            if (state.tick.planets[from].population < size) {
+                return { error: "Invalid size! You don't have enough troops." };
+            }
+            if (fromTo.has(from.toString() + "_" + to.toString())) { // TODO: toString is probably slow
+                return { error: "Invalid step! You can't send troops from one planet to another more than once." };
+            }
+            fromTo.add(from.toString() + "_" + to.toString()); // Little bit ugly
+            troops.push({ playerID, from, to, size });
+        }
+        return troops;
+    }
+    catch(e) {
+        console.log(e);
+        return {error: "Invalid input"}; // TODO: better error message
+    }
+}
+
+function updateState(state: GameState, steps: UserStep[]): GameState {
+    // Add troops that are leaving their planets
+    for (let step of steps){
+        for (let troop of step) {
+            state.tick.planets[troop.from].population -= troop.size;
+            state.tick.troops.push({
+                id: troopIDCounter++,
+                from: troop.from,
+                to: troop.to,
+                player: troop.playerID,
+                size: troop.size,
+                endTick: state.tick.id + state.planetsDistances[troop.from][troop.to],
+            });
+        }
+    }
+
+    // Process troops that are arriving to their planets, preparing for fight
+    let planetWaitingList: {who: PlayerID, size: number}[][] = Array.from(Array(state.planets.length), () => []); // TODO: use dictionary instead of array
+    for (let i = 0; i < state.tick.troops.length; i++) {
+        if (state.tick.troops[i].endTick === state.tick.id) {
+            let planet = state.tick.troops[i].to;
+            let user = { who: state.tick.troops[i].player, size: state.tick.troops[i].size };
+            // Add planets to
+            planetWaitingList[planet].push(user);
+            console.log("Arrived:", planetWaitingList);
+            state.tick.troops.splice(i, 1); // Removing troop from list
```

Comment:
Ha a tick.troops egy endTick szerint indexelt Map lenne, akkor
1. itt egy lépésben ki tudnád törölni az összeset aki ebben a körben érkezett meg  O(n^2) helyett
2. a for ciklusban sem kellene az összes troopot végignézni